### PR TITLE
SONAR-8857 support organizations in QProfileRef

### PR DESCRIPTION
**Pull request by simonbrandhof**
SONAR-8857 support organizations in QProfileRef

 + new method QProfileWsSupport#getProfile(DbSession,QProfileRef)

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRef.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+   * @return non-null name
+   * @throws IllegalStateException if {@link #hasKey()} does not return {@code false}
+   */
+  public String getName() {
+    checkState(name != null, "Name is not present. Please call hasKey().");
+    return name;
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRef.java]: MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1142)

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java [hunk 2]
```diff
@@ -160,6 +160,10 @@ private void setDefault(DbSession session, QualityProfileDto profile) {
     db.qualityProfileDao().update(session, profile.setDefault(true));
   }
 
+  /**
+   * @deprecated replaced by {@link org.sonar.server.qualityprofile.ws.QProfileWsSupport#getProfile(DbSession, org.sonar.server.qualityprofile.ws.QProfileRef)}
+   */
+  @Deprecated
   public QualityProfileDto find(DbSession dbSession, QProfileRef ref) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileRef.java [hunk 3]
```diff
@@ -36,7 +36,10 @@
  *   <li>by its key</li>
  *   <li>by the couple {language, name}</li>
  * </ul>
+ *
+ * @deprecated replaced by {@link org.sonar.server.qualityprofile.ws.QProfileRef}
  */
+@Deprecated
 public class QProfileRef {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileRef.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QProfileFactory.java#L182](https://github.com/SonarSource/sonarqube/blob/3561779274f3694447a6552fd6d168185ee0ae5c/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L182): Remove this use of "selectByNameAndLanguage"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![MINOR][MINOR] [QProfileFactory.java#L196](https://github.com/SonarSource/sonarqube/blob/3561779274f3694447a6552fd6d168185ee0ae5c/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L196): Remove this use of "selectByNameAndLanguages"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![MINOR][MINOR] [QProfileFactory.java#L217](https://github.com/SonarSource/sonarqube/blob/3561779274f3694447a6552fd6d168185ee0ae5c/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L217): Remove this use of "selectByNameAndLanguage"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![INFO][INFO] [QProfileWsSupport.java#L86](https://github.com/SonarSource/sonarqube/blob/3561779274f3694447a6552fd6d168185ee0ae5c/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileWsSupport.java#L86): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRef.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+/**
+ * Reference to a Quality profile as defined by requests to web services api/qualityprofiles.
+ * The two exclusive options to reference a profile are:
+ * <ul>
+ *   <li>by its id (to be deprecated)</li>
+ *   <li>by the tuple {organizationKey, language, name}</li>
+ * </ul>
+ */
+public class QProfileRef {
```
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRef.java]: I do not like abbreviations. I would suggest `QualityProfileReference` as a name.

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileRefTest.java [hunk 5]
```diff
@@ -38,47 +37,60 @@
...
 
   @Test
-  public void getName_throws_ISE_if_key_ref() {
+  public void getName_throws_ISE_on_reference_by_key() {
     QProfileRef ref = QProfileRef.fromKey("foo");
 
     expectedException.expect(IllegalStateException.class);
+    expectedException.expectMessage("Name is not present. Please call hasKey().");
     ref.getName();
   }
```
Comment (bartfastiel) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileRefTest.java]: `getOrganization_throws_ISE_if_key_ref` seems to be missing.

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileRefTest.java [hunk 6]
```diff
@@ -151,10 +178,10 @@ public void test_equals_and_hashCode_of_key_ref() {
...
-    QProfileRef name1bis = QProfileRef.fromName("js", "one");
-    QProfileRef name2 = QProfileRef.fromName("js", "two");
-    QProfileRef name1OtherLang = QProfileRef.fromName("java", "one");
+    QProfileRef name1 = QProfileRef.fromName("org1", "js", "one");
+    QProfileRef name1bis = QProfileRef.fromName("org1", "js", "one");
+    QProfileRef name2 = QProfileRef.fromName("org1", "js", "two");
+    QProfileRef name1OtherLang = QProfileRef.fromName("org1", "java", "one");
     QProfileRef key = QProfileRef.fromKey("one");
 
     assertThat(name1.equals(name1)).isTrue();
```
Comment (bartfastiel) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileRefTest.java]: Concerning line `197`:

Use `org.sonar.server.language.LanguageTesting#newLanguage(java.lang.String)` instead.

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileWsSupport.java [hunk 7]
```diff
@@ -86,4 +87,20 @@ public OrganizationDto getDefaultOrganization(DbSession dbSession) {
...
+   * Get the Quality profile specified by the reference {@code ref}.
+   *
+   * @throws org.sonar.server.exceptions.NotFoundException if the specified organization or profile do not exist
+   */
+  public QualityProfileDto getProfile(DbSession dbSession, QProfileRef ref) {
+    QualityProfileDto profile;
+    if (ref.hasKey()) {
+      profile = dbClient.qualityProfileDao().selectByKey(dbSession, ref.getKey());
+    } else {
+      OrganizationDto org = getOrganizationByKey(dbSession, ref.getOrganizationKey().orElse(null));
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileWsSupport.java]: No, `getOrganizationByKey()` aims to support that.

**Pull request closed without merging**